tpm2/tpm2_test.go

```diff
@@ -0,0 +1,318 @@
+		uint32(0x00010001),
+		[]byte(nil),
+	}
+	privateBlob, publicBlob, err := CreateKey(rw, parentHandle, []int{7}, "01020304", "01020304", keyparms)
+	if err != nil {
+		t.Fatalf("CreateKey failed: %s", err)
+	}
+
+	keyHandle, _, err := Load(rw, parentHandle, "", "01020304", publicBlob, privateBlob)
+	if err != nil {
+		t.Fatalf("Load failed: %s", err)
+	}
+
+	_, _, _, err = ReadPublic(rw, keyHandle)
+	if err != nil {
+		t.Fatalf("ReadPublic failed: %s", err)
+	}
+
+	if err = FlushContext(rw, keyHandle); err != nil {
+		t.Fatalf("FlushContext failed: %s", err)
+	}
+	if err = FlushContext(rw, parentHandle); err != nil {
+		t.Fatalf("FlushContext failed: %s", err)
+	}
+}
+
+func TestCombinedEndorsementTest(t *testing.T) {
+	if !*runIntegration {
+		t.SkipNow()
+	}
+
+	rw, err := OpenTPM("/dev/tpm0")
+	if err != nil {
+		t.Fatalf("OpenTPM failed: %s", err)
+	}
+	defer rw.Close()
+
+	primaryparms := RSAParams{
+		AlgRSA,
+		AlgSHA1,
+		0x00030072,
+		[]byte(nil),
+		AlgAES,
+		128,
+		AlgCFB,
+		AlgNull,
+		0,
+		2048,
+		uint32(0x00010001),
+		[]byte(nil),
+	}
+	parentHandle, publicBlob, err := CreatePrimary(rw, HandleOwner, []int{0x7}, "", "", primaryparms)
+	if err != nil {
+		t.Fatalf("CreatePrimary failed: %s", err)
+	}
+
+	keyparms := RSAParams{
+		AlgRSA,
+		AlgSHA1,
+		0x00030072,
+		[]byte(nil),
+		AlgAES,
+		128,
+		AlgCFB,
+		AlgNull,
+		0,
+		2048,
+		uint32(0x00010001),
+		[]byte(nil),
+	}
+	privateBlob, publicBlob, err := CreateKey(rw, parentHandle, []int{7}, "", "01020304", keyparms)
+	if err != nil {
+		t.Fatalf("CreateKey failed: %s", err)
+	}
+
+	keyHandle, _, err := Load(rw, parentHandle, "", "", publicBlob, privateBlob)
+	if err != nil {
+		t.Fatalf("Load failed: %s", err)
+	}
+
+	_, name, _, err := ReadPublic(rw, keyHandle)
+	if err != nil {
+		t.Fatalf("ReadPublic failed: %s", err)
+	}
+
+	// Generate Credential
+	credential := []byte{1, 2, 3, 4, 5, 6, 7, 8, 9, 0xa, 0xb, 0xc, 0xd, 0xe, 0xf, 0x10}
+	credBlob, encryptedSecret0, err := MakeCredential(rw, parentHandle, credential, name)
+	if err != nil {
+		FlushContext(rw, keyHandle)
+		FlushContext(rw, parentHandle)
+		t.Fatalf("MakeCredential failed: %s", err)
+	}
+
+	recoveredCredential1, err := ActivateCredential(rw, keyHandle, parentHandle, "01020304", "", credBlob, encryptedSecret0)
+	if err != nil {
+		FlushContext(rw, keyHandle)
+		FlushContext(rw, parentHandle)
+		t.Fatalf("ActivateCredential failed: %s", err)
+	}
+	if bytes.Compare(credential, recoveredCredential1) != 0 {
+		FlushContext(rw, keyHandle)
+		FlushContext(rw, parentHandle)
+		t.Fatalf("Credential and recovered credential differ: got %v, want %v", recoveredCredential1, credential)
+	}
+
+	FlushContext(rw, keyHandle)
+}
+
+func TestCombinedContextTest(t *testing.T) {
+	if !*runIntegration {
+		t.SkipNow()
+	}
+	rw, err := OpenTPM("/dev/tpm0")
+	if err != nil {
+		t.Fatalf("OpenTPM failed: %v", err)
+	}
+	defer rw.Close()
+
+	pcrs := []int{7}
+	keySize := 2048
+	quotePassword := ""
+
+	primaryparms := RSAParams{
```

Comment:
If most of the parameters are the same across all uses of RSAParams in this test file, consider making a template version at the top of this file and filling in only the parts that need to change for each test.

Reply:
Done.